File: palindrome.rb
def is_palindrome(s)
  chars = s.downcase.scan(/[a-z0-9]/)
  left_index = 0
  right_index = chars.length - 1

  while left_index < right_index do
    return false if chars[left_index] != chars[right_index]
    left_index += 1
    right_index -= 1
  end

  true
end

p is_palindrome('A man, a plan, a canal: Panama')
p is_palindrome('')
p is_palindrome('a.')

# Alternate solution

def is_palindrome(s)
  return true if s.length < 2

  left = 0
  right = s.length - 1

  while !alphanumeric?(s[left])
    left += 1
  end

  while !alphanumeric?(s[right])
    right -= 1
  end

  while left < right
    return false if s[left].downcase != s[right].downcase

    left += 1
    right -= 1

    while !alphanumeric?(s[left])
      left += 1
    end

    while !alphanumeric?(s[right])
      right -= 1
    end
  end

  true
end

def alphanumeric?(char)
  !!char.match(/[a-zA-Z0-9]/)
end

File: palindrome.js
// Given a string, determine if it is a palindrome, considering only alphanumeric
// characters and ignoring cases.
// 
// For example,
// "A man, a plan, a canal: Panama" is a palindrome.
// "race a car" is not a palindrome.
// 
// Note:
// Have you consider that the string might be empty? This is a good question 
// to ask during an interview.
// 
// For the purpose of this problem, we define empty string as valid palindrome.
//
// ========================
//
// if string length < 2, return true
//
// Two pointers
//   left at beginning, right at end
//     increment left until it is at alphanumeric char
//     decrement right until it is at alphanumeric char
//   while left < right
//     return false if lowercase value at left != lowercase value at right
//     increment left until it is at alphanumeric char
//     decrement right until it is at alphanumeric char
//   return true

var isPalindrome = function(s) {
  if (s.length < 2) return true;
  var left = 0;
  var right = s.length - 1;

  while (!isAlphanumeric(s[left]) && left < right) {
    left++;
  }
  while (!isAlphanumeric(s[right]) && left < right) {
    right--;
  }

  while (left < right) {
    if (s[left].toLowerCase() !== s[right].toLowerCase()) {
      return false;
    }

    left++;
    right--;

    while (!isAlphanumeric(s[left])) {
      left++;
    }
    while (!isAlphanumeric(s[right])) {
      right--;
    }
  }

  return true;
};

var isAlphanumeric = function(s) {
  var charCode = s.charCodeAt(0);

  return charCode >= 48 && charCode <= 57 ||
         charCode >= 65 && charCode <= 90 ||
         charCode >= 97 && charCode <= 122;
};

